p2p/pex/addrbook.go

```diff
@@ -295,6 +307,87 @@ func (a *addrBook) GetSelection() []*p2p.NetAddress {
 	return allAddr[:numAddresses]
 }
 
+// GetSelectionWithBias implements AddrBook.
+// It randomly selects some addresses (old & new). Suitable for peer-exchange protocols.
+//
+// Each address is picked randomly from an old or new bucket according to the
+// biasTowardsNewAddrs argument, which must be between [0, 100] (or else is truncated to
+// that range) and determines how biased we are to pick an address from a new
+// bucket.
+func (a *addrBook) GetSelectionWithBias(biasTowardsNewAddrs int) []*p2p.NetAddress {
+	a.mtx.Lock()
+	defer a.mtx.Unlock()
+
+	if a.size() == 0 {
+		return nil
+	}
+
+	if biasTowardsNewAddrs > 100 {
+		biasTowardsNewAddrs = 100
+	}
+	if biasTowardsNewAddrs < 0 {
+		biasTowardsNewAddrs = 0
+	}
+
+	numAddresses := cmn.MaxInt(
+		cmn.MinInt(minGetSelection, a.size()),
+		a.size()*getSelectionPercent/100)
+	numAddresses = cmn.MinInt(maxGetSelection, numAddresses)
+
+	selection := make([]*p2p.NetAddress, numAddresses)
+
+	oldBucketAndIndex := make(map[int]int)
+	var oldIndex int
+	newBucketAndIndex := make(map[int]int)
+	var newIndex int
+	var bucket map[string]*knownAddress
+
+	i := 0
+ADDRS_LOOP:
+	for i < numAddresses {
+		pickFromOldBucket := int((float64(i)/float64(numAddresses))*100) >= biasTowardsNewAddrs
+		bucket = make(map[string]*knownAddress)
+
+		// loop until we pick a random non-empty bucket
+		for len(bucket) == 0 {
```

Comment:
Won't `bucket` be always of length zero given we construct a new map on line 349?

Reply:
no, because later we do `bucket = a.bucketsXXX[newIndex]`